metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/sql/DirectSqlGetPartition.java

```diff
@@ -187,35 +205,100 @@ public int getPartitionCount(
-                        values = Maps.newHashMap();
-                        values.put(FIELD_DATE_CREATED, createdDate + "");
-                    }
-                    if (Strings.isNullOrEmpty(filterExpression)
-                        || filter.evaluatePartitionExpression(filterExpression, name, uri, isBatched, values)) {
-                        names.add(name);
-                    }
+                    uris.add(rs.getString(PARTITION_URI));
                 }
-                return names;
+                return uris;
             };
-            result = getHandlerResults(tableName.getDatabaseName(),
-                tableName.getTableName(), filterExpression, partitionNames,
-                SQL.SQL_GET_PARTITIONS_WITH_KEY_URI, handler, sort, pageable);
+            result = getHandlerResults(tableName.getDatabaseName(), tableName.getTableName(),
+                null, partitionNames, SQL.SQL_GET_PARTITIONS_URI, handler, sort, pageable);
+        }
+        this.fastServiceMetric.recordTimer(
+            HiveMetrics.TagGetPartitionKeys.getMetricName(), registry.clock().wallTime() - start);
+        return result;
+    }
+
+
+    /**
+     * query partitions using filters from name or uri column.
+     */
+    private List<String> filterPartitionsColumn(
+        final String databaseName,
+        final String tableName,
+        final List<String> partitionNames,
+        final String columnName,
+        final String filterExpression,
+        final Sort sort,
+        final Pageable pageable) {
+        final FilterPartition filter = new FilterPartition();
+        // batch exists
+        final boolean isBatched =
+            !Strings.isNullOrEmpty(filterExpression) && filterExpression.contains(FIELD_BATCHID);
+        final boolean hasDateCreated =
+            !Strings.isNullOrEmpty(filterExpression) && filterExpression.contains(FIELD_DATE_CREATED);
+
+        ResultSetExtractor<List<String>> handler = rs -> {
+            final List<String> columns = Lists.newArrayList();
+            while (rs.next()) {
+                final String name = rs.getString(PARTITION_NAME);
+                final String uri = rs.getString(PARTITION_URI);
+                final long createdDate = rs.getLong(FIELD_DATE_CREATED);
+                Map<String, String> values = null;
+                if (hasDateCreated) {
+                    values = Maps.newHashMap();
+                    values.put(FIELD_DATE_CREATED, createdDate + "");
+                }
+                if (Strings.isNullOrEmpty(filterExpression)
+                    || filter.evaluatePartitionExpression(filterExpression, name, uri, isBatched, values)) {
+                    columns.add(rs.getString(columnName));
+                }
+            }
+            return columns;
+        };
+        return getHandlerResults(databaseName,
+            tableName, filterExpression, partitionNames,
+            SQL.SQL_GET_PARTITIONS_WITH_KEY_URI, handler, sort, pageable);
```

Comment:
We should have a different query to get partition keys and a different one to get uris.

Reply:
But key and uri are needed in the sql to support filter, even though we return one of them.